File: src/main/java/io/github/divyeshmangla/timetable/parser/BatchExtractor.java
package io.github.divyeshmangla.timetable.parser;

import io.github.divyeshmangla.timetable.excel.CellUtils;
import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;

import java.util.Comparator;
import java.util.Map;
import java.util.TreeMap;
import java.util.regex.Pattern;

public class BatchExtractor {
    private static final Pattern BATCH_REGEX = Pattern.compile("^(\\d[A-Z]\\d[A-Z])$");
    private static final Comparator<String> BATCH_COMPARATOR = (key1, key2) -> {
        int num1 = extractNumber(key1);
        int num2 = extractNumber(key2);
        int result = Integer.compare(num1, num2);
        return result != 0 ? result : key1.compareTo(key2);
    };

    private final Sheet sheet;

    public BatchExtractor(Sheet sheet) {
        this.sheet = sheet;
    }

    public Map<String, Cell> extract() {
        Cell dayCell = CellUtils.findCellInFirstColumn(sheet, "day");
        if (dayCell == null) {
            throw new IllegalStateException("Could not find 'day' cell in the first column");
        }

        Map<String, Cell> batches = new TreeMap<>(BATCH_COMPARATOR);
        Row row = dayCell.getRow();

        for (int i = 0; i < row.getLastCellNum(); i++) {
            Cell cell = row.getCell(i);
            if (cell == null) continue;

            String cellContents = cell.toString().trim();
            if (cellContents.isEmpty()) continue;

            if (BATCH_REGEX.matcher(cellContents).matches()) {
                String normalized = normalizeBatchName(cellContents);
                batches.put(normalized, cell);
            }
        }

        return batches;
    }

    private static int extractNumber(String batchName) {
        return Integer.parseInt(batchName.substring(2));
    }

    private String normalizeBatchName(String rawName) {
        if (rawName == null || rawName.length() != 4) {
            return rawName;
        }

        char lastLetter = rawName.charAt(3);
        int position = lastLetter - 'A' + 1;

        return rawName.substring(0, 3) + position;
    }
}

File: internal/parser/extractor/batch_extractor.go
package extractor

import (
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/DivyeshMangla/tiet-timetable/internal/parser"
	"github.com/DivyeshMangla/tiet-timetable/internal/types"
	"github.com/xuri/excelize/v2"
)

var batchRegex = regexp.MustCompile(`^\d[A-Z]\d[A-Z]$`)

type BatchExtractor struct {
	file      *excelize.File
	sheetName string
}

func NewBatchExtractor(file *excelize.File, sheetName string) *BatchExtractor {
	return &BatchExtractor{
		file:      file,
		sheetName: sheetName,
	}
}

func (be *BatchExtractor) Extract() (map[types.BatchID]parser.CellLocation, error) {
	dayRow, _, found := parser.FindCellInFirstColumn(be.file, be.sheetName, "day")
	if !found {
		return nil, fmt.Errorf("could not find 'day' cell in the first column")
	}

	rows, err := be.file.GetRows(be.sheetName)
	if err != nil {
		return nil, fmt.Errorf("failed to get rows: %w", err)
	}

	if dayRow >= len(rows) {
		return nil, fmt.Errorf("day row out of bounds")
	}

	batches := make(map[types.BatchID]parser.CellLocation)
	dayRowData := rows[dayRow]

	for col := 0; col < len(dayRowData); col++ {
		cellValue := strings.TrimSpace(dayRowData[col])
		if cellValue == "" {
			continue
		}

		if batchRegex.MatchString(cellValue) {
			normalized := normalizeBatchName(cellValue)
			batches[types.BatchID(normalized)] = parser.CellLocation{
				Row: dayRow,
				Col: col,
			}
		}
	}

	return sortBatches(batches), nil
}

func extractNumber(batchName string) int {
	if len(batchName) < 3 {
		return 0
	}
	num, _ := strconv.Atoi(batchName[2:3])
	return num
}

func normalizeBatchName(rawName string) string {
	if len(rawName) != 4 {
		return rawName
	}

	lastLetter := rawName[3]
	position := int(lastLetter-'A') + 1

	return rawName[:3] + strconv.Itoa(position)
}

func sortBatches(batches map[types.BatchID]parser.CellLocation) map[types.BatchID]parser.CellLocation {
	type batchEntry struct {
		id       types.BatchID
		location parser.CellLocation
	}

	entries := make([]batchEntry, 0, len(batches))
	for id, loc := range batches {
		entries = append(entries, batchEntry{id: id, location: loc})
	}

	sort.Slice(entries, func(i, j int) bool {
		num1 := extractNumber(string(entries[i].id))
		num2 := extractNumber(string(entries[j].id))
		if num1 != num2 {
			return num1 < num2
		}
		return string(entries[i].id) < string(entries[j].id)
	})

	sorted := make(map[types.BatchID]parser.CellLocation, len(entries))
	for _, entry := range entries {
		sorted[entry.id] = entry.location
	}

	return sorted
}
